Added LSpace viewer that spawns an L-Space viewer as a separate process.

Britefury opened the pull request:
This PR adds a 'View in LSpace' item to the View menu.

It spawns the LSpace 'eco_viewer' example program to display the content of the currently visible document.

snim2 commented:
`RuntimeError: Could not execute LSpace viewer at /home/snim2/kcl/lspace/target/release/examples/eco_viewer, set its path using the LSPACE_VIEWER_PATH environment variable`

Maybe worth adding the `LSPACE_VIEWER_PATH` to the new settings pane that @ptersilie recently committed?

Britefury commented:
Okay, I'll go do that :)

snim2 commented:
Also setting that env var to be the root of a clone of the lspace repo does not work, which surprised me a little!

Britefury commented:
Good point, thanks! :)

Britefury commented:
Done; now uses the settings and the path provided is the repo root rather than the executable root as per @snim2's suggestion.

Britefury commented:
I have modified the LSpace viewer to account for changes to the Rust side.

Prior to these changes, Eco described what it wanted in terms of a document being composed of lines, which were composed of tokens, which were textual items; similar in some ways to the structure of an Eco document.

The new system allows a generic presentation to be described. It is serialised to JSON and sent to the Rust app for rendering. As a result, we can now display whatever we like. The presentation API is a taster of the API that we expect to develop in the future.

snim2 commented:
OK, currently blocked on Issue 29 in the LSpace repo: https://github.com/softdevteam/lspace/issues/29

snim2 commented on lib/eco/lspace_ext/lspace.py:
```diff
@@ -0,0 +1,126 @@
+    Start the LSpace eco_viewer external tool
+
+    :param pres: a `Pres` instance to show or another object that will be coerced into a `Pres`.
+    """
```
I think `:param` is syntax for one of the Python Javadoc equivalents. I don't _think_ we are using this elsewhere in Eco, so probably best to rewrite this in prose.

Britefury replied:
Fixed.

snim2 commented on lib/eco/lspace_ext/view_ecodoc_in_lspace.py:
```diff
@@ -0,0 +1,91 @@
+    View the content of the given `TreeManager` in the LSpace eco_viewer tool
+    :param tree_manager: a `TreeManager` instance
+    :return:
+    """
```
likewise.

Britefury replied:
Fixed.

snim2 commented on lib/eco/lspace_ext/lspace.py:
```diff
@@ -0,0 +1,126 @@
+
+        out, err = proc.communicate()
+        print out
+        if len(err) > 0:
```
Why print to STDOUT?

Britefury replied:
Okay, an exception is raised if errors are detected. Otherwise the output is printed to STDOUT as it can show one or two interesting things.

snim2 replied:
OK, but this is something that you will need but the user will not, I think? So, I would suggest not committing it. Alternatively, we could add a global `--debug` flat on the command line, for things like this.
/cc @ptersilie

snim2 commented on lib/eco/lspace_ext/lspace.py:
```diff
@@ -0,0 +1,126 @@
+
+        try:
+            proc = subprocess.Popen([lspace_viewer_path, a_path],
+                                    cwd=lspace_viewer_dir, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```
OK, so this is where you actually launch the external viewer. I guess the GUI window is actually provided by the Rust code. At present, you are running the LSpace viewer modally, so the viewer blocks the main GUI thread in Eco, and in Ubuntu the main window is greyed out a little (because it cannot accept events, etc). I'm not sure whether this is a good thing or not. If you take the view that an end user should not be able to edit their document or open another Eco tool while looking at the LSpace viewer, then this code is fine. If you think the user should be able to keep the LSpace viewer open and still use Eco, then you need to change this code a bit. The LSpace viewer would need to run in its own thread or process (e.g. with http://pyqt.sourceforge.net/Docs/PyQt4/qthread.html).

snim2 commented on lib/eco/lspace_ext/lspace.py:
```diff
@@ -0,0 +1,126 @@
+                "{0}, you can set the LSpace root using either the settings "
+                "dialog or the LSPACE_ROOT environment variable".format(
+                lspace_viewer_path))
+
```
this formatting looks a bit odd to me. Might be worth looking at how string substitution is formatted elsewhere in Eco (I think perhaps we don't use `format()`?)

ptersilie commented on lib/eco/lspace_ext/lspace.py:
```diff
@@ -0,0 +1,126 @@
+        '~', 'kcl', 'lspace'))
+
+
+def _valid_lspace_root(lspace_root=None):
```
I think @ltratt  and I agreed to only have one path pointing to external stuff to no confuse the user with multiple settings. In Eco's case we decided to only use the settings and not use environment vars at all.

Quote from https://github.com/softdevteam/eco/pull/65:

> I think it's probably best to remove the environment stuff, otherwise people might define the variable outside Eco, then define it inside, and be confused as to which actually effects the outcome.

ptersilie commented:
Okay looks good to me. Maybe some squashing and then I'm ready to merge this.

Britefury commented:
Okay, will do.

Commit:
Implemented LSpace viewer that spawns LSpace in a separate process.
The viewer is composed of two parts.
1. a generic viewer launcher that allows a generic presentation to be described using and API that approximates the final intended API
2. a Eco document viewer that uses the aforementioned.

Britefury commented:
Squashed.

snim2 commented:
@ptersilie do you want to assign yourself to this? I don't mind either way

ptersilie commented:
@snim2 Ah, didn't know you were assigned to this.
